### ISSUE TITLE
Wrong output values in TFLite_Detection_PostProcess operation when max_classes_per_detection > 1

### ISSUE THREAD

**Issue by defr8001** [tensorflow/tensorflow#47551]
**System information**

OS Platform and Distribution: Ubuntu 18.04
TensorFlow installed from: pypi
TensorFlow version: 1.15.2
Python version: 3.6.9

**Describe the current behavior**

When you add the NMS operation at exporting the TFlite graph with the option `max_classes_per_detection > 1`, you'll get more than one class per detection as expected, but the class IDs and their corresponding scores are not consecutive in the output array. Instead, the output arrays have a position without value, that only contains garbage memory.

For example, with `max_classes_per_detection = 2`, you get:

output[0] = detection1_top1_class_id
output[1] = garbage
output[2] = garbage
output[3] = detection1_top2_class_id
output[4] = garbage
output[5] = garbage
...

**Describe the expected behavior**

output[0] = detection1_top1_class_id
output[1] = detection1_top2_class_id
output[2] = detection2_top1_class_id
...

**Standalone code to reproduce the issue**

Export the model with:

```bash
python3 export_tflite_ssd_graph.py \
    --pipeline_config_path=configs/ssdlite_mobiledet_cpu_320x320_coco_sync_4x4.config \
    --trained_checkpoint_prefix=ssdlite/model.ckpt-100000 \
    --output_directory=ssdlite/tflite \
    --max_detections=10 \
    --max_classes_per_detection=2 \
    --add_postprocessing_op=true
```

```
# Load TFLite model and allocate tensors.
interpreter = tf.lite.Interpreter(model_path='ssdlite/tflite/model.tflite'))
interpreter.allocate_tensors()

# Get input and output tensors.
input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()

# Test model
input_shape = input_details[0]['shape']
input_data = load_image_into_numpy_array(image_paths[0])
interpreter.set_tensor(input_details[0]['index'], input_data)

interpreter.invoke()

# The function `get_tensor()` returns a copy of the tensor data
count = int(interpreter.get_tensor(output_details[3]['index'])[0])
boxes = interpreter.get_tensor(output_details[0]['index'])[:, :count]
classes = interpreter.get_tensor(output_details[1]['index'])[:, :count]
scores = interpreter.get_tensor(output_details[2]['index'])[:, :count]

# classes[1] and scores[1] will contain random numbers from unassigned memory slots
```

I think that the error stems from this line of code:

https://github.com/tensorflow/tensorflow/blob/127adf71b4a995265cdcc960de38c164c93d53a9/tensorflow/lite/kernels/detection_postprocess.cc#L710

`box_offset` should have the value of `output_box_index` (which is already incremented in the inner loop where the top classes of the detection are assigned) or move the line `output_box_index++;` out of the inner loop to really make it account for the number of detections

**Comment by abattery** [comment]
@thaink could you review this suggestion?

**Comment by thaink** [comment]
Agreed that that loop looks wrong. Basically, we can do either stack all valid values first like:
output[0] = detection1_top1_class_id
output[1] = detection2_top1_class_id
output[2] = detection2_top2_class_id
output[3] = garbage (because detection 1 has only one class)
Or use memory like for a matrix:
output[0] = detection1_top1_class_id
output[1] = garbage (because detection 1 has only one class)
output[2] = detection2_top1_class_id
output[3] = detection2_top2_class_id
We are following the second pattern (so people can random access to a detection). But the implementation looks wrong.

In my idea, the box_offset should be: ` int box_offset = max_categories_per_anchor * output_box_index + col; `
and the `output_box_index++;` should be moved to outer loop.

**Comment by defr8001** [comment]
Thanks @thaink and @abattery,

I agree with your idea to solve the problem. If the bug is fixed, would it be solved both in TF1 and TF2?

**Comment by thaink** [comment]
It is fixed at commit 0b2560e217cf5bc7c9f931a3185c9c8e3f2e9d5e in master branch.

**Comment by defr8001** [comment]
Awesome, thanks!
